nifi-nar-bundles/nifi-mqtt-bundle/nifi-mqtt-processors/src/main/java/org/apache/nifi/processors/mqtt/ConsumeMQTT.java

```diff
@@ -0,0 +1,341 @@
+                throw new ProcessException("Attempting to stop processor but stored ProcessSessionFactory is not set, there are messages in the MQTT Queue and it is configured to " +
+                        "finish processing the messages.");
+            }
+        }
+    }
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) throws ProcessException {
+        if (mqttQueue.isEmpty() && !isConnected()){
+            logger.info("Queue is empty and client is not connected. Attempting to reconnect.");
+
+            try {
+                reconnect();
+            } catch (MqttException e) {
+                logger.error("Connection to " + broker + " lost (or was never connected) and ontrigger connect failed. Yielding processor", e);
+                context.yield();
+            }
+        }
+
+        if (mqttQueue.isEmpty()) {
+            return;
+        }
+
+        transferQueue(session);
+    }
+
+    private void transferQueue(ProcessSession session){
+        while (!mqttQueue.isEmpty()) {
+            FlowFile messageFlowfile = session.create();
+            final MQTTQueueMessage mqttMessage = mqttQueue.peek();
+
+            Map<String, String> attrs = new HashMap<>();
+            attrs.put(BROKER_ATTRIBUTE_KEY, broker);
+            attrs.put(TOPIC_ATTRIBUTE_KEY, mqttMessage.getTopic());
+            attrs.put(QOS_ATTRIBUTE_KEY, String.valueOf(mqttMessage.getQos()));
+            attrs.put(IS_DUPLICATE_ATTRIBUTE_KEY, String.valueOf(mqttMessage.isDuplicate()));
+            attrs.put(IS_RETAINED_ATTRIBUTE_KEY, String.valueOf(mqttMessage.isRetained()));
+
+            messageFlowfile = session.putAllAttributes(messageFlowfile, attrs);
+
+            messageFlowfile = session.write(messageFlowfile, new OutputStreamCallback() {
+                @Override
+                public void process(final OutputStream out) throws IOException {
+                    out.write(mqttMessage.getPayload());
+                }
+            });
+
+            String transitUri = new StringBuilder(broker).append(mqttMessage.getTopic()).toString();
+            session.getProvenanceReporter().receive(messageFlowfile, transitUri);
+            session.transfer(messageFlowfile, REL_MESSAGE);
+            mqttQueue.remove(mqttMessage);
+            session.commit();
+        }
+    }
+
+    private class ConsumeMQTTCallback implements MqttCallback {
+
+        @Override
+        public void connectionLost(Throwable cause) {
+            logger.warn("Connection to " + broker + " lost", cause);
+            try {
+                reconnect();
+            } catch (MqttException e) {
+                logger.error("Connection to " + broker + " lost and callback re-connect failed.");
+            }
+        }
+
+        @Override
+        public void messageArrived(String topic, MqttMessage message) throws Exception {
+            logger.info("MQTT message arrived on topic:" + topic);
```

Comment:
General comment (applies to all logger callbacks) (except may be for _error_); While logger itself will handle the logging at the pre-configured level it is generally recommended to still wrap it in _IF_ bloc for cases where message String is composed (e.g., "foo" + this.getBar() + " bar") as you are essentially go through the process of building a message string that may never see the light of day.